Disable .numpy() and .tolist() for tensor subclasses subclasses and fix .tolist() for conjugated and negated tensors

anjali411 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #66082

Fixes https://github.com/pytorch/pytorch/issues/66024 #65779 

cc @ezyang @anjali411 @dylanbespalko @mruberry @Lezcano @nikitaved @albanD

Differential Revision: [D31615588](https://our.internmc.facebook.com/intern/diff/D31615588)

anjali411 committed:
Disable .numpy() and .tolist() for tensor subclassesnsor subclasses and fix .tolist() for conjugated and negated tensors

[ghstack-poisoned]

anjali411 committed:
Update on "Disable .numpy() and .tolist() for tensor subclassesnsor subclasses and fix .tolist() for conjugated and negated tensors"


Fixes https://github.com/pytorch/pytorch/issues/66024 #65779 

[ghstack-poisoned]

albanD commented on test/test_python_dispatch.py:
```diff
@@ -460,6 +460,14 @@ def test_nested_enable_python_mode(self) -> None:
+        x = LoggingTensor(torch.tensor([2.0, 3.0], requires_grad=True))
+        with self.assertRaisesRegex(RuntimeError, "is not supported for tensor subclasses."):
+            x.tolist()
+            x.numpy() 
```
You need to separate this one in a different context manager, otherwise, it is never tested.

albanD commented on test/test_python_dispatch.py:
```diff
@@ -460,6 +460,14 @@ def test_nested_enable_python_mode(self) -> None:
+            x.tolist()
+            x.numpy() 
+        with self.assertRaisesRegex(RuntimeError, "is not supported for tensor subclasses."):
+            self.assertEqual(x, None)
```
Should we open an issue here to remove the use of tolist() in the assert equal?

ezyang commented:
This seems like an OK hotfix but I personally prefer something like https://github.com/pytorch/pytorch/issues/66024#issuecomment-933924798 for the long term

I'll let Alban do the final approve as he has substantive comments here.

anjali411 committed:
Update on "Disable .numpy() and .tolist() for tensor subclasses subclasses and fix .tolist() for conjugated and negated tensors"


Fixes https://github.com/pytorch/pytorch/issues/66024 #65779 

cc ezyang anjali411 dylanbespalko mruberry @Lezcano nikitaved albanD

[ghstack-poisoned]

anjali411 committed:
Update on "Disable .numpy() and .tolist() for tensor subclasses subclasses and fix .tolist() for conjugated and negated tensors"


Fixes https://github.com/pytorch/pytorch/issues/66024 #65779 

cc ezyang anjali411 dylanbespalko mruberry @Lezcano nikitaved albanD

[ghstack-poisoned]

albanD reviewed (approved):
Only one test fix and then it is good to go!

anjali411 committed:
Update on "Disable .numpy() and .tolist() for tensor subclasses subclasses and fix .tolist() for conjugated and negated tensors"


Fixes https://github.com/pytorch/pytorch/issues/66024 #65779 

cc ezyang anjali411 dylanbespalko mruberry @Lezcano nikitaved albanD

[ghstack-poisoned]

albanD reviewed (approved):
final rev

anjali411 committed:
Update on "Disable .numpy() and .tolist() for tensor subclasses subclasses and fix .tolist() for conjugated and negated tensors"


Fixes https://github.com/pytorch/pytorch/issues/66024 #65779 

cc ezyang anjali411 dylanbespalko mruberry @Lezcano nikitaved albanD

Differential Revision: [D31615588](https://our.internmc.facebook.com/intern/diff/D31615588)

[ghstack-poisoned]

anjali411 commented:
@anjali411 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31615588).
